src/ai/backend/common/message_queue/redis_queue/queue.py

```diff
@@ -0,0 +1,185 @@
+from typing import Any, AsyncGenerator, Mapping, Optional, Self, override
+
+from ai.backend.common.types import RedisTarget
+
+from ..abc import AbstractAnycaster, AbstractBroadcaster, AbstractConsumer, AbstractSubscriber
+from ..abc.queue import AbstractMessageQueue
+from ..types import BroadcastMessage, BroadcastPayload, MessageId, MQMessage
+from .anycaster import RedisAnycaster
+from .broadcaster import RedisBroadcaster
+from .consumer import RedisConsumer, RedisConsumerArgs
+from .subscriber import RedisSubscriber
+
+_DEFAULT_AUTOCLAIM_IDLE_TIMEOUT = 300_000  # 5 minutes
+
+
+@dataclass
+class RedisMQArgs:
+    # Required arguments
+    anycast_stream_key: str
+    broadcast_channel: str
+    consume_stream_keys: Optional[set[str]]
+    subscribe_channels: Optional[set[str]]
+    group_name: str
+    node_id: str
+    db: int
+    # Optional arguments
+    autoclaim_idle_timeout: int = _DEFAULT_AUTOCLAIM_IDLE_TIMEOUT
+    autoclaim_start_id: Optional[str] = None
+
+
+class RedisQueue(AbstractMessageQueue):
+    """
+    Redis message queue implementation with composable components.
+
+    This queue can contain any combination of the four components:
+    - Anycaster: for point-to-point messaging
+    - Broadcaster: for pub/sub broadcasting
+    - Consumer: for consuming messages from streams
+    - Subscriber: for receiving broadcast messages
+    """
+
+    _anycaster: AbstractAnycaster
+    _broadcaster: AbstractBroadcaster
+    _consumer: AbstractConsumer
+    _subscriber: AbstractSubscriber
+
+    def __init__(
+        self,
+        anycaster: AbstractAnycaster,
+        broadcaster: AbstractBroadcaster,
+        consumer: AbstractConsumer,
+        subscriber: AbstractSubscriber,
+    ) -> None:
+        self._anycaster = anycaster
+        self._broadcaster = broadcaster
+        self._consumer = consumer
+        self._subscriber = subscriber
+
+    @classmethod
+    async def create(cls, redis_target: RedisTarget, args: RedisMQArgs) -> Self:
+        """
+        Create a RedisQueue instance with all four components.
+        """
+        # Create anycaster
+        anycaster = await RedisAnycaster.create(redis_target, args.anycast_stream_key, args.db)
+
+        # Create broadcaster
+        broadcaster = await RedisBroadcaster.create(redis_target, args.broadcast_channel, args.db)
+
+        # Create consumer
+        consume_stream_keys = args.consume_stream_keys or set()
+        consumer = await RedisConsumer.create(
+            redis_target,
+            RedisConsumerArgs(
+                consume_stream_keys,
+                args.group_name,
+                args.node_id,
+                args.db,
+                args.autoclaim_idle_timeout,
+                args.autoclaim_start_id,
+            ),
+        )
+
+        # Create subscriber
+        subscribe_channels = args.subscribe_channels or set()
+        subscriber = await RedisSubscriber.create(redis_target, subscribe_channels, args.db)
```

Comment:
I'd suggest creating a `RedisSubscriber` only if `subscribe_channels` is not empty.
ref: https://github.com/valkey-io/valkey-glide/blob/1d6f82e50817f39715bc10fb49fc11aaf46ffab0/python/glide-async/python/glide/glide_client.py#L529-L532